README
OpenTorpedo - What is OpenTorpedo?
==================================

OpenTorpedo is a submarine simulation (subsim) which tries to emulate
the behaviour of modern submarines. The player is placed in command
of a submarine and is able to move about in a deep ocean environment.

Combat missions are now being added to OpenTorpedo, and ships with
sonar will run from torpedoes and, if they can detect the player,
will return fire.



Why a subsim?
=============

I've always been a fan of subsims and played commercial games,
but couldn't find a modern subsim for open source operating systems
(such as Linux and FreeBSD). Danger from the Deep is available,
but it's stictly for World War II u-boats, and has relatively high
graphics requirements. Looking around I found a half dozen or so
subsims that had begun and dropped away shortly after beginning development.

The LinuxSSN project had laid a solid foundation before the developers
dropped the project and I decided to pick up where they left off.
The heart of OpenTorpedo comes from their project and attempts to further
their work.



Licensing info
==============

The OpenTorpedo project is licensed under the GNU General Public License,
version 3. The LinuxSSN project was also GNU GPL licensed and we
carry on that tradition. Copyright Jesse Smith <jessefrgsmith@yahoo.ca>
Please see the LICENSE file for further details.

Data files, missions, artwork and images included in OpenTorpedo are
licensed under the Creative Commons CC0 1.0 Universal license. 
Please see the ART-LICENSE file for full details.



What do I need to build OpenTorpedo?
====================================

To build OpenTorpedo you'll need a C++ compiler (GNU's Compiler Collection
is recommended) and the following development libraries:

SDL
SDL_gfx
SDL_image
SDL_mixer

Assuming the above libraries are installed, it should be possible to
build OpenTorpedo with the following commands

tar zxf OpenTorpedo-1.4.tar.gz
cd OpenTorpedo
make



Running OpenTorpedo
===================

Once OpenTorpedo has built, an executable file (called OpenTorpedo) will
be placed in the top level directory. From there
the game can be run with the command:

./OpenTorpedo

The following command line parameters are recognized:

-w		Run in windowed mode, this is the default
-f		Run in full screen
-s		Enable sound effects
-v		Display version number and exit
-h		Display help and exit
-m <mission> 	Load a specific mission



Control Keys
============

The player's submarine can be controlled with either the mouse 
or keyboard commands. The following key commands are recognized.

Pilot the submarine..............
Plus key		Increase sub's speed
Minus key		Slow down
Left keypad (4)		Turn left
Right keypad (6)	Turn right
Down keypad (2)		Dive deeper
Up keypad (8)		Rise toward the surface
Q			Go up one thermal
A			Hold depth here
Z			Go down one thermal
S			Surface
X			Go to periscope depth


System controls
F1			Sonar screen
F2			Map screen
F3			Weapons station
F4			Helm console
F5			ESM signals screen
F6			Radar screen
F10			Take screen shot
TAB			Select next target
Delete			Increase map scale
Insert			Decrease map scale
Up, down, left, right	Manipulate map
P			Pause the game
Page Up			Compress Time
Page Down		Slow time
ESC 			Quit


Weapon controls 
T			Load a torpedo
N			Load a noise maker
U			Unload a tube
F			Fire a tube

Each of the weapon station keys should be followed by
a number indication which tube should be loaded or fired.
The tubes are numbered 1-6.


Sonar controls
G			Send active sonar pulse.
E			Extend sonar array
R			Retract towed sonar array
W			Hold towed sonar in place



Contact information
===================

Did you find a bug? Would you like to request a feature? Did
you patch something and want to share it? Send an e-mail to
jessefrgsmith@yahoo.ca. Please place "OpenTorpedo" in the subject line.



CHANGELOG
  - Fixed compile error on Windows, missing header include.

======= OpenTorpedo-1.4 ==================

  - Added new mission, provided by David Luckie.

====== OpenTorpedo-1.3 ===================

  - Added unistd.h and corrected DrawPixel declaration
    to fix compile/build errors.

  - Removed un-needed -k command line option. It
    was buggy and causing compiler warnings.


========= OpenTorpedo-1.2 Jun 3, 2012 ===========

  - Applied makefile patch from Vincent.

  - We no longer create font and largeont files in
    the directory "./data/" OpenTorpedo used to try to
    save our fonts in case the font files were missing,
    but this is no longer required and is probably
    improper behaviour. This fix closes Debian bug
    #675744.
    Thanks to Guillem for pointing out the error.



========== OpenTorpedo-1.1 Nov 7, 2011 ============

  - Updated makefile to support more data locations.

  - Helicopters will guard friendly ships if they cannot
    hear enemies.

  - Subs attack a little slower to avoid alerting everyone.

  - Helicopters no longer rapid-fire torpedoes.

  - Subs will now stay below a thermal if they hear helicopters.

  - Helicopters will shoot at enemy subs.

  - Helicopters should now track enemy submarines.

  - Helicopter now moves randomly when searching for
    targets and it has nothing better to do.

  - Added debug compiler flag DEBUG_HELICOPTER.

  - Removed "ships" index variable. With linked lists we
    should not need it.

  - Added helicopter class.

======== OpenTorpedo-1.0 Oct 22, 2011 ============

  - Game now recognizes OPTDIR variable, for when OpenTorpedo
    is installed in /opt/OpenTorpedo/

  - Corrected/improved makefile.
    Patch provided by Vincent.

  - Added optinstall option for Makefile, installs
    OpenTorpedo under /opt/OpenTorpedo

========== OpenTorpedo-0.9 September 26, 2011  =========

  - Added sub vs sub mission.

  - Stack holding ESM data is now of dynamic size. Should no
    longer crash when faced with many surface targets.

  - Updated makefile. Added manual page, desktop entry and icon.
    Files and updates provided by Vincent on behave of 
    the Debian project.

  - Vessels in attack mood who have no target slow down
    to cruising speed so they can hear better.

  - Added escort mission.

  - When an escorted ship is sunk the mission fails.

  - Fixed bug which prevented all sounds from showing up on
    sonar display.

  - We should now be able to make missions where a vessel's
    goal is to stay alive for N seconds.

  - Sound signals should show up brighter on sonar display.

  - Added new recon mission.

  - AI subs should not "randomly" move to surface during
    patrol.

  - Vessel depth changes a little faster now, even at
    slow speeds. Simplified depth change calculation.

  - When traveling in reverse the sub should rise/fall in
    the proper direction.

  - AI subs should change depth sometimes.

  - Fixed bug which would prevent AI from shooting at
    near targets.

  - AI vessels should shoot at different ranges depending
    on their mood and a little randomness.

  - Enemy surface ships now detect radar signals and
    investigate.

  - Pinging at escorts in a convoy will cause the escorts to
    examine the pinging vessel.

  - AI now notices pinging and will move to investigate.

  - Added keys for moving to periscope depth and surfacing.
    X and S respectively.

  - Radar now gives us the location, but not a firm
    contact on surface vessels. Prevents "cheating'
    when on recon missions.

  - Ship type is now included in target information.

  - Added new carrier class ships.

  - AI vessels in ATTACK mood should now chase after the
    origin of torpedoes.

  - Reduced thermal dampening of sound from 25% to 15%

  - Player should be informed if they have been hit.

  - Updated English PDF manual.

  - Added makefile target "tarball" for creating source packages.

  - Stack object is now a linked list, not an array. Allows for
    more ships/subs/sonar data.

============ OpenTorpedo-0.8 released September 18, 2011 ==========

  - Display text feedback when changing depth or holding depth.

  - Added new mission, hunt for a Typhoon.

  - We should no longer display target information if we
    do not have a strong contact.
    Re-wrote Next_Target function to only return strong targets.

  - Added controls to go up/down a thermal. Added key to hold
    current depth. See README file for details.

  - Depth and thermal meter now appears on helm screen and nav screen.

  - Use SDL_gfx functions to draw arrows on helm screen.
    Minor code clean-up. Patch provided by Rick.

  - Added patch from Rick to show ordered heading vs actual
    heading on helm screen.

  - Added depth and thermal indicator to the tactical map screen.

  - Added Spanish translation of the manual from Fernando Liebana.

  - Make sure we define DATADIR, even if we don't use a makefile.

  - Thermal layers now filter sound.

  - Rewrote code for getting command line options to make it
    compatable across operating systems.

  - Compatibility functions for WIN32 created. These functions
    are stored in winfunctions.cpp and winfunctions.h
    strncasecmp and strcasestr functions are now duplicated
    and wrapped up in ifdefs.

  - Converted snprintf calls to sprintf for WIN32

  - Converted fabs() calls to abs() in radar.cpp for the WIN32 platform.

  - Dependency unistd.h removed from main.cpp

  - Added thermal layers to listening function Can_Hear().

  - Added DEBUGMAP flag in Makefile to help test MAP class.

  - Added MAP class, which will handle depth and thermal layer information.

============ OpenTorpedo-0.7 released August 15, 2011 =============

  - Added sounds for pinging, torpedo launch and explosions.

  - Added sounds directory creation to makefile.

  - Added click button for active sonar.

  - Torpedo room buttons looks more torpedo-like.


src/main.h
/***************************************************************************
                          main.h  -  description
                             -------------------
    begin                : Tue Sep 19 2000
    copyright            : (C) 2000 by Michael Bridak
    email                : michael.bridak@verizon.net
$Id: main.h,v 1.11 2003/07/18 03:50:00 mbridak Exp $
 ***************************************************************************/

/***************************************************************************

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License.
     .

 ***************************************************************************/

#ifndef MAIN_H
#define MAIN_H

#ifndef VERSION
#define VERSION 1.4
#endif

#include "helicopter.h"

// delay between game loops
#define GAME_DELAY 10
#define MAP_FACTOR 20
#define MAX_MAP_SCALE 500

#define PERISCOPE_DEPTH 50
#define MAX_SUBS 20

#define DEFAULT_SEA_STATE 3
#define SHIP_HEIGHT 100


/** Our main program and control loop
  *@author Michael Bridak
  */
// Some defs for easy living.
#define		SONAR	1
#define		NAVMAP	2
#define		TMA	3
#define		QUIT	4
#define		COMPRESSTIME	5
#define		UNCOMPRESSTIME	6
#define		INCREASESPEED	7
#define		DECREASESPEED	8
#define		DUMPSCREEN	9
#define		TURNPORT	10
#define		TURNSTARBOARD	11
#define		INCREASEDEPTH	12
#define		DECREASEDEPTH	13
#define		INCREASEMAPSCALE	14
#define		DECREASEMAPSCALE	15
#define		CENTERDISPLAY	16
#define		EXTENDARRAY	17
#define		RETRACTARRAY	18
#define		STOPWINCH	19
#define		PAUSEGAME	20
#define		SHIPCONTROL 21
#define		SCROLLMAPUP	22
#define		SCROLLMAPDOWN	23
#define		SCROLLMAPLEFT	24
#define		SCROLLMAPRIGHT	25
#define		TOGGLESPHERICALTOWED 26
#define		TOGGLETRUERELATIVE 27
#define		UPPERCRTBUTTON	28
#define		LOWERCRTBUTTON	29
#define		ASSIGNTRACKER 30
#define		TRACKER1 31
#define		TRACKER2 32
#define		TRACKER3 33
#define		TRACKER4 34
#define		ESM 35
#define		RADAR 36

// Radar Event Updates
#define		TOGGLER10	37
#define		TOGGLER20	38
#define		TOGGLER30	39
#define		TOGGLER40	40
#define		TOGGLER50	41
#define		TOGGLER60	42
#define		RADARDOWN	43
#define		RADARUP		44
#define		RING0	        45
#define		RING5	        46
#define		RING10   	47
#define		RING15  	48
#define		RING20  	49
#define		RING25  	50

// Esm Event Updates
#define		ESMDOWN		51
#define		ESMUP		52

// Control Station Updates
#define		ASTOP	        53
#define		A13	        54
#define		A23      	55
#define		ASTD     	56
#define		AFULL    	57
#define		AFLK            58
#define		B13     	59
#define		B23      	60
#define		BSTD     	61
#define		BEMER      	62
#define		DESIGNATECONTACT	63
#define 	SWITCHTARGET 64
#define 	WEAPONS 65
#define		LOAD_TORPEDO 66
#define		LOAD_NOISEMAKER 67
#define		UNLOAD_TUBE 68
#define		FIRE_TUBE 69
#define		WHICHTUBE 70
#define		USE_TUBE 71
#define		CUTARRAY 72
#define         SEND_PING 73
#define         UP_THERMAL 74
#define         HOLD_DEPTH 75
#define         DOWN_THERMAL 76
#define		GO_PERISCOPE_DEPTH 77
#define		GO_SURFACE 78


//globals

// int	ships = 19; // why does ships equal 19, isn't this set in CreateShips?
float	flowandambientnoise;
SDL_Surface	*screen=NULL; //The screen we are goin to be looking at
SDL_Surface	*towedarrayscreen=NULL; //scratch blit surface

// rdm
SDL_Surface	*radarscreen=NULL; //scratch blit surface
SDL_Surface	*esmscreen=NULL; //scratch blit surface
SDL_Surface *controlscreen=NULL;



SDL_Surface	*tempscreen=NULL; //scratch screen
SDL_Surface	*sonarbuttonup=NULL; //button element
SDL_Surface	*sonarbuttondown=NULL; //button element
SDL_Surface	*navbuttonup=NULL; //button element
SDL_Surface	*navbuttondown=NULL; //button element
SDL_Surface	*tmabuttonup=NULL; //button element
SDL_Surface	*tmabuttondown=NULL; //button element
SDL_Surface	*shipcontrolbuttonup=NULL;//button element
SDL_Surface	*shipcontrolbuttondown=NULL;//button element
SDL_Surface	*esmbuttonup=NULL;//button element
SDL_Surface	*esmbuttondown=NULL;//button element
SDL_Surface	*radarbuttonup=NULL;//button element
SDL_Surface	*radarbuttondown=NULL;//button element
SDL_Surface	*quitbuttonup=NULL; //button element
SDL_Surface	*quitbuttondown=NULL; //button element
SDL_Surface	*plusbuttonup=NULL; //button element
SDL_Surface	*plusbuttondown=NULL; //button element
SDL_Surface	*centerbuttonup=NULL; //button element
SDL_Surface	*centerbuttondown=NULL; //button element
SDL_Surface	*leftbuttonup=NULL; //button element
SDL_Surface	*leftbuttondown=NULL; //button element
SDL_Surface	*rightbuttonup=NULL; //button element
SDL_Surface	*rightbuttondown=NULL; //button element
SDL_Surface	*upbuttonup=NULL; //button element
SDL_Surface	*upbuttondown=NULL; //button element
SDL_Surface	*downbuttonup=NULL; //button element
SDL_Surface	*downbuttondown=NULL; //button element
SDL_Surface	*minusbuttonup=NULL; //button element
SDL_Surface	*minusbuttondown=NULL; //button element
SDL_Surface	*ncscale=NULL;
SDL_Surface	*scscale=NULL;
SDL_Surface	*truerel[2]={NULL, NULL};
SDL_Surface	*sphertowed[2]={NULL, NULL};
SDL_Surface	*tb16winchon=NULL;
SDL_Surface	*tb16winchoff=NULL;
SDL_Surface	*extendtb16[2]={NULL, NULL};
SDL_Surface	*retracttb16[2]={NULL, NULL};
SDL_Surface	*uppercrtoff=NULL;
SDL_Surface	*uppercrton=NULL;
SDL_Surface	*lowercrtoff=NULL;
SDL_Surface	*lowercrton=NULL;
SDL_Surface	*assigntrackerwidget[2]={NULL, NULL};
SDL_Surface	*tracker1[2]={NULL, NULL};
SDL_Surface	*tracker2[2]={NULL, NULL};
SDL_Surface	*tracker3[2]={NULL, NULL};
SDL_Surface	*tracker4[2]={NULL, NULL};
SDL_Surface	*noisemaker_image = NULL;
SDL_Surface	*torpedo_image = NULL;


//a mutex so we don't crash and burn
//when accessing the same globals
//in a callback function.
SDL_mutex *tmamutex;

SDL_Rect destination_rectangle; //rectangle structure for blitting

DFont fnt(const char*, const char *);
DFont fnt2(const char*, const char *);

int tube_action = 0;    // what are we doing with a torpedo tube
int tube_to_use = -1;   // the tube to perform this action on

int drawsonar; //flags for UpdateDisplay()
int drawmap; //flags for UpdateDisplay()
int drawtma; //flags for UpdateDisplay()
//rdm
int drawweapons;  // display weapons screen
int drawradar; //flags for UpdateDisplay()
int drawesm; //flags for UpdateDisplay()
int drawcontrol; //flags for UpdateDisplay

int mapscale = 25;
int mapcenter = 1;   // by default, center the map on our ship
int timecompression;
int sonarwidget, navwidget, quitwidget, tmawidget; //button pressed flags
int shipcontrolwidget, esmwidget, radarwidget; //shouldn't these be bools?
Uint32 textcolor, black, white, red, green, yellow, grey, mapcolor;//Place to hold color info
Uint32 dark_green, dark_red, brown, dark_grey;
bool show_spherical_traces = true, northcenter = true, pause_game = false;
bool assigntracker = false;
Submarine *current_target = NULL;
int update_weapons_screen = TRUE;
int my_mission_status;
int should_update_everything = TRUE;

GameClock Clock;
// Submarine Subs[MAX_SUBS]; //Somewhere to put our little floating buddies.
Submarine *Subs = NULL;   // all subs, ships 
Submarine *player = NULL;
Contact Contacts[MAX_SUBS]; //Workable contacts.
Submarine *torpedoes;   // a linked-list of torpedoes
Helicopter *helicopters;  // a linked-list of helicopters
TowedArray TB16;
TargetMotionAnalysis Tma;
msg Message;
AnBqq5 SonarStation(Subs, TB16, Tma, Message);


Radar RadarStation(Subs);
Esm EsmStation(Subs);
Control ControlStation(Subs);


//Danger  I don't know what I'm doing here
struct position {
	float lat;
	float lon;
};
std::list<position> SubPosLog;  //STL incompetence DANGER!
//End Danger


//declarations
void	SetupScreen(bool);
void	CreateShips(int mission_number, MAP *map);
void	UpdateSensors(void);
void	UpdateDisplay(void);
void	TakeCommands(void);
double	RelativeBearing(Submarine *observer, Submarine *target);
int	ReciprocalBearing(int bearing);
double	AngleOnBow(Submarine *observer, Submarine *target);
int	maximize360 ( int course );
int	minimize360 ( int course );
double	CalculateRange(Submarine *observer, Submarine *target);
void	LatLonDifference(Submarine *observer, Submarine *target, double *platdif, double *plondif);
void	PositionCursor(int, int);
void	ClearScreen(void);
void    Display_Target(void);
void	DrawMap(void);
void    Draw_Depth_Meter(Submarine *my_sub, int which_screen);
/*
void	DrawPixel(SDL_Surface *screen, int x, int y, Uint32 Color);
void	DrawLine(SDL_Surface *screen, int X1, int Y1, int X2, int Y2, Uint32 Color);
void	DrawArc(SDL_Surface *screen, int X1, int Y1, int Radius, int Theta1, int Theta2, Uint32 Color);
void	DrawCircle(SDL_Surface *screen, int X1, int Y1, int Radius, Uint32 Color);
void	DrawDiamond(SDL_Surface *screen, int X1, int Y1, int Size, char Direction, Uint32 Color);
void	DrawBox(SDL_Surface *screen, int X1, int Y1, int Size, char Direction, Uint32 Color);
void    DrawRectangle(SDL_Surface *screen, int x1, int y1, int x2, int y2, Uint32 Color);
void	DrawCross(SDL_Surface *screen, int X1, int Y1, int Size, char Direction, Uint32 Color);
*/
void	PlaceShips(int, int = 0, int = 0, Submarine *target=NULL); // the int = 0's are default values
// inline void	DrawPixel(SDL_Surface *screen, int x, int y, Uint32 Color);
// inline void	DrawLine(SDL_Surface *screen, int X1, int Y1, int X2, int Y2, Uint32 Color);
void	LoadScreen(int screen_to_load);
void	ShowStation(int station);
void	DisplaySonar();
void	DisplayTMA(int xoffset=0, int yoffset=0);
void    DisplayWeapons(void);
void	LoadWidgets(void);
void	UnLoadWidgets(void);
void	DisplayWidgets(void);
void	DisplayNavigationWidgets();
void	DisplayTMAWidgets();
void	DisplayESMWidgets();
void	DisplayRADARWidgets();
void	ResetWidgetFlags(void);
inline int	RandInt(int to);
void	MapIcon(int x, int y, int ShipType, int Friend, Uint32 color);
void	DirectionalPointer(int X, int Y, int Heading, int speed, Uint32 Color);
int	InBaffles(Submarine *observer, Submarine *target, int sensor);
void	SoundEnvironment(void);
float	Any_Detection(double Range, Submarine *observer, Submarine *target);
float	Radar_Detection(double Range, Submarine *observer, Submarine *target);
float	Esm_Detection(double Range, Submarine *observer, Submarine *target);
// float	Sonar_Detection(double Range, int ObserverShipId, int TargetShipId);
float   Sonar_Detection_New(double Range, Submarine *observer, Submarine *target);
inline int		Clamp(int);
inline double	Clamp(double);
Uint32	TmaTimer(Uint32 interval, void *param);
Uint32	timerfunc(Uint32 interval, void *param);

Submarine *Add_Ship(Submarine *all_torpedoes, Submarine *new_torpedo);
Submarine *Remove_Ship(Submarine *all_torpedoes, Submarine *old_torpedo);
void Remove_Inactive_Ship(Submarine *victim);  // find and remove inactive ship
Helicopter *Add_Helicopter(Helicopter *all_helicopters, Helicopter *new_helicopter);
Helicopter *Remove_Helicopter(Helicopter *all_helicopters, Helicopter*old_helicopter);

#endif

